### PR TITLE
chore(stdlib): Add docs to List module

### PR DESCRIPTION
**Pull request by phated**
While prepping for release, I noticed that the List module didn't have any graindoc added to it. With the automated docs updates on the website, this would wipe out our List documentation.

I took most of this from the Array module and updated it for the differences. I also noticed some inconsistencies in Array that I'll PR after this.

Please thoroughly review the docs for clarity and consistency (especially with our other docs).

### CODE REVIEW

**Comment by phated** [comment]
Any changes made here during review should also be reflected in #1299

File: stdlib/list.gr [hunk 1]
```diff
@@ -46,42 +95,122 @@ let reverse = list => {
...
+ *
+ * @since v0.2.0
+ * @history v0.1.0: Originally named `foldLeft`
+ * @history v0.2.0: Renamed to `reduce`
+ */
+export let rec reduce = (fn, initial, list) => {
   match (list) {
-    [] => acc,
-    [first, ...rest] => reduce(fn, fn(acc, first), rest),
+    [] => initial,
```
Comment (marcusroberts) [inline stdlib/list.gr:154]: Maybe just me but acc for accumulator makes sense, initial makes me think it's the initial (unchanging) value

Comment (marcusroberts) [reply stdlib/list.gr:154]: Changed my mind, this makes the documentation look much clearer

**Review by marcusroberts — APPROVED** [review]
This all read really well to me, and reading through it I learnt a lot about the rich functionality of List (and some of the performance issues that come from lists!)

Great work putting all this together

**Review by peblair — APPROVED** [review]
Looks great! Thanks for putting this together!

File: stdlib/list.gr [hunk 2]
```diff
@@ -120,7 +289,16 @@ let rec forEach = (fn, list) => {
   }
 }
 
-let forEachi = (fn, list) => {
+/**
+ * Iterates a list, calling an iterator function with each element.
```
[the remark this answers is not in the document]
Comment (phated) [reply stdlib/list.gr:293]:
```suggestion
 * Iterates a list, calling an iterator function on each element.
```

**Review by ospencer — APPROVED** [review]
Fantastic!

**Pull request merged**